### ISSUE TITLE
Documents and Revision Trees

### ISSUE THREAD

**Issue by bawelter** [fluttercouch/fluttercouch#21]
There is an issue with the approach of how documents are handled.  Currently documents are essentially recreated because the document is created as a new document every time and overwrites the current document.  So every time a document is saved and another one exists there is a conflict and the last write wins strategy resolves the conflict.

The issue with this is that the revision tree is destroyed each time a document is saved and may cause bandwidth issues when syncing documents if the entire document needs synced across all clients.  Need to look into this and find possible solutions.

**Comment by oltrenuovefrontiere** [comment]
I suppose that internally the Document object given by the Couchbase SDK has an internal reference to the revision number. This number in used by the sdk when a new version is saved into the database.

If this is the case, I think that storing that revision number in the dart implementation too can be a far better approach.

**Comment by bawelter** [comment]
I would agree on that since there is no way to initialize a document with the revision number.

**Comment by bawelter** [comment]
When saving we need to get the document on the platform side and compare revisions.  If they are different then we need to throw back a conflict with the conflicting document to the dart code.  If they are the same then we will apply changes to the doc and only save if there isn't a conflict on save, if there is then we will throw back a conflict with the conflicting document.

That was the strategy I had in mind anyway, but I won't be able to get to this any time soon.

**Comment by oltrenuovefrontiere** [comment]
We can use [getSequence()](https://docs.couchbase.com/mobile/2.5.0/couchbase-lite-java/com/couchbase/lite/Document.html#getSequence--) to compare "versions" of the same document. If we pass back to dart the sequence number of the document the user bases his modifications on, we can allow saving only if the sequence number of the database document is still the same. But we have to store in some way the changes made on the document in the dart side to efficiently update the native document.

In Couchbase Lite 2.5 there isn't a direct conflict resolution handling (i.e., the last save wins or it fails on conflicts). I have minimal experience with those type of errors. How do you handle them in native SDKs?

**Comment by bawelter** [comment]
I was looking into this and I think we add a keepAlive function and do our own garbage collection on this.  If you call the keepAlive function then you must dispose of the object yourself.  I say this because blobs will also give us issues especially with memory issues. So blobs would need to be fetched over the channel every time from the originating object because they are technically stored in the object.

**Comment by oltrenuovefrontiere** [comment]
I've just read the announcement of 2.6 release, and wondering if custom conflict resolution could come in handy to resolve our issue https://docs.couchbase.com/couchbase-lite/2.6/index.html#custom-conflict-resolution
